File: Complementar - Estruturas de dados (Didático)/Algoritmos34/Alg34.php
<?php
/*
♦ Algoritmo 34: Estrutura de Dados Não Linear Dinâmica (não Estática) de Árvore Binária: Árvore AVL.
*/


class No {
    public $valor;
    public $esquerda, $direita;
    public $altura;

    public function __construct($valor) {
        $this->valor = $valor;
        $this->esquerda = $this->direita = null;
        $this->altura = 1;
    }
}

class ArvoreAVL {
    private $raiz;

    public function __construct() {
        $this->raiz = null;
    }

    private function altura($no) {
        if ($no == null)
            return 0;
        return $no->altura;
    }

    private function balanceamento($no) {
        if ($no == null)
            return 0;
        return $this->altura($no->esquerda) - $this->altura($no->direita);
    }

    private function rotacaoDireita($y) {
        $x = $y->esquerda;
        $T2 = $x->direita;

        $x->direita = $y;
        $y->esquerda = $T2;

        $y->altura = max($this->altura($y->esquerda), $this->altura($y->direita)) + 1;
        $x->altura = max($this->altura($x->esquerda), $this->altura($x->direita)) + 1;

        return $x;
    }

    private function rotacaoEsquerda($x) {
        $y = $x->direita;
        $T2 = $y->esquerda;

        $y->esquerda = $x;
        $x->direita = $T2;

        $x->altura = max($this->altura($x->esquerda), $this->altura($x->direita)) + 1;
        $y->altura = max($this->altura($y->esquerda), $this->altura($y->direita)) + 1;

        return $y;
    }

    public function inserir($valor) {
        $this->raiz = $this->inserirRecursivo($this->raiz, $valor);
    }

    private function inserirRecursivo($no, $valor) {
        if ($no == null)
            return new No($valor);

        if ($valor < $no->valor)
            $no->esquerda = $this->inserirRecursivo($no->esquerda, $valor);
        else if ($valor > $no->valor)
            $no->direita = $this->inserirRecursivo($no->direita, $valor);
        else
            return $no;

        $no->altura = 1 + max($this->altura($no->esquerda), $this->altura($no->direita));

        $balanceamento = $this->balanceamento($no);

        if ($balanceamento > 1 && $valor < $no->esquerda->valor)
            return $this->rotacaoDireita($no);

        if ($balanceamento < -1 && $valor > $no->direita->valor)
            return $this->rotacaoEsquerda($no);

        if ($balanceamento > 1 && $valor > $no->esquerda->valor) {
            $no->esquerda = $this->rotacaoEsquerda($no->esquerda);
            return $this->rotacaoDireita($no);
        }

        if ($balanceamento < -1 && $valor < $no->direita->valor) {
            $no->direita = $this->rotacaoDireita($no->direita);
            return $this->rotacaoEsquerda($no);
        }

        return $no;
    }

    private function imprimirEmOrdemRecursivo($no) {
        if ($no != null) {
            $this->imprimirEmOrdemRecursivo($no->esquerda);
            echo $no->valor . " ";
            $this->imprimirEmOrdemRecursivo($no->direita);
        }
    }

    public function imprimirEmOrdem() {
        $this->imprimirEmOrdemRecursivo($this->raiz);
    }
}

// Teste
$arvore = new ArvoreAVL();
$arvore->inserir(10);
$arvore->inserir(5);
$arvore->inserir(20);
$arvore->inserir(3);
$arvore->inserir(8);
$arvore->inserir(15);

echo "Em Ordem:\n";
$arvore->imprimirEmOrdem();

?>

File: Complementar - Estruturas de dados (Didático)/Algoritmos34/Alg34.py
'''
♦ Algoritmo 34: Estrutura de Dados Não Linear Dinâmica (não Estática) de Árvore Binária: Árvore AVL.
'''


class No:
    def __init__(self, valor):
        self.valor = valor
        self.esquerda = None
        self.direita = None
        self.altura = 1

class ArvoreAVL:
    def __init__(self):
        self.raiz = None

    def altura(self, no):
        if no is None:
            return 0
        return no.altura

    def balanceamento(self, no):
        if no is None:
            return 0
        return self.altura(no.esquerda) - self.altura(no.direita)

    def rotacao_direita(self, y):
        x = y.esquerda
        T2 = x.direita

        x.direita = y
        y.esquerda = T2

        y.altura = max(self.altura(y.esquerda), self.altura(y.direita)) + 1
        x.altura = max(self.altura(x.esquerda), self.altura(x.direita)) + 1

        return x

    def rotacao_esquerda(self, x):
        y = x.direita
        T2 = y.esquerda

        y.esquerda = x
        x.direita = T2

        x.altura = max(self.altura(x.esquerda), self.altura(x.direita)) + 1
        y.altura = max(self.altura(y.esquerda), self.altura(y.direita)) + 1

        return y

    def inserir(self, valor):
        self.raiz = self.inserir_recursivo(self.raiz, valor)

    def inserir_recursivo(self, no, valor):
        if no is None:
            return No(valor)

        if valor < no.valor:
            no.esquerda = self.inserir_recursivo(no.esquerda, valor)
        elif valor > no.valor:
            no.direita = self.inserir_recursivo(no.direita, valor)
        else:
            return no

        no.altura = 1 + max(self.altura(no.esquerda), self.altura(no.direita))

        balanceamento = self.balanceamento(no)

        if balanceamento > 1 and valor < no.esquerda.valor:
            return self.rotacao_direita(no)

        if balanceamento < -1 and valor > no.direita.valor:
            return self.rotacao_esquerda(no)

        if balanceamento > 1 and valor > no.esquerda.valor:
            no.esquerda = self.rotacao_esquerda(no.esquerda)
            return self.rotacao_direita(no)

        if balanceamento < -1 and valor < no.direita.valor:
            no.direita = self.rotacao_direita(no.direita)
            return self.rotacao_esquerda(no)

        return no

    def imprimir_em_ordem_recursivo(self, no):
        if no is not None:
            self.imprimir_em_ordem_recursivo(no.esquerda)
            print(no.valor, end=" ")
            self.imprimir_em_ordem_recursivo(no.direita)

    def imprimir_em_ordem(self):
        self.imprimir_em_ordem_recursivo(self.raiz)


# Teste
arvore = ArvoreAVL()
arvore.inserir(10)
arvore.inserir(5)
arvore.inserir(20)
arvore.inserir(3)
arvore.inserir(8)
arvore.inserir(15)

print("Em Ordem:")
arvore.imprimir_em_ordem()
